### PR TITLE
SystemCapability Subscription

### PR DESCRIPTION
**Pull request by ShobhitAd**
#0167 Implement subscription for GetSystemCapabilities RPC. Used to inform the App Service Consumer of newly published services

This PR is **ready** for review.

### Risk
This PR makes **major** API changes.

### Testing Plan
Create an ATF test script to test SystemCapabilities Subscription

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/on_system_capability_updated_notification.cc [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+                              hmi_capabilities,
+                              policy_handler) {}
+
+OnSystemCapabilityUpdatedNotification::
+    ~OnSystemCapabilityUpdatedNotification() {}
+
+void OnSystemCapabilityUpdatedNotification::Run() {
+  LOG4CXX_AUTO_TRACE(logger_);
+
+  // TODO, figure out how to handle unknown enums for SystemCapabilityType
```
Comment (JackLivio) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/on_system_capability_updated_notification.cc]: I dont think core will have to handle unknown enums in SystemCapabilityType. Future unknown app services will be covered under the `APP_SERVICES` system capability type. This assumption needs clarification though. I believe the servicesSupported parameter in appServiceCapabilities should be changed to type string.

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/extensions/system_capability_app_extension.cc [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                   << system_capability_type);
+  auto it = subscribed_data_.find(system_capability_type);
+  if (it != subscribed_data_.end()) {
+    subscribed_data_.erase(it);
+    return true;
+  }
+  return false;
+}
+
+void SystemCapabilityAppExtension::unsubscribeFrom() {
```
Comment (JackLivio) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/extensions/system_capability_app_extension.cc]: Maybe change name to `unsubscribeFromAll` for clarity.

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/sdl_rpc_plugin.cc [hunk 3]
```diff
@@ -32,41 +32,57 @@
...
+  } else if (plugins::ApplicationEvent::kDeleteApplicationData == event) {
+    DeleteSubscriptions(application);
+  }
+}
+
+void SDLRPCPlugin::DeleteSubscriptions(app_mngr::ApplicationSharedPtr app) {
+  auto& ext = SystemCapabilityAppExtension::ExtractExtension(*app);
+  auto subscriptions = ext.Subscriptions();
+  for (auto& sub : subscriptions) {
+    ext.unsubscribeFrom(sub);
```
Comment (JackLivio) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/sdl_rpc_plugin.cc]: Is it necessary to loop through the subscriptions when there is a function that does unsubscribeFromAll()?

File: src/components/application_manager/src/app_service_manager.cc [hunk 4]
```diff
@@ -68,6 +68,37 @@ smart_objects::SmartObject AppServiceManager::PublishAppService(
   service_record[strings::service_active] = true;
   published_services_.insert(std::pair<std::string, smart_objects::SmartObject>(
       service_id, service_record));
+
+  smart_objects::SmartObjectSPtr notification =
```
Comment (JackLivio) [inline src/components/application_manager/src/app_service_manager.cc]: Either in the appServiceManager, or the application app service extension, should have a service record object that stores of the service records for all published app services. This might need to be a separate task on its own PR

File: src/components/interfaces/MOBILE_API.xml [hunk 5]
```diff
@@ -3973,8 +3957,68 @@
         <param name="weatherServiceData" type="WeatherServiceData" mandatory="false"/>
     </struct>
 
+    <enum name="ServiceUpdateReason">
```
Comment (JackLivio) [inline src/components/interfaces/MOBILE_API.xml]: Since 5.1

File: src/components/interfaces/MOBILE_API.xml [hunk 6]
```diff
@@ -6847,6 +6891,9 @@
         <param name="systemCapabilityType" type="SystemCapabilityType" mandatory="true">
             <description>The type of system capability to get more information on</description>
         </param>
+        <param name="subscribe" type="Boolean" mandatory="false">
```
Comment (JackLivio) [inline src/components/interfaces/MOBILE_API.xml]: since 5.1

**Comment by JackLivio** [comment]
Reminder note to add unit tests before merging pr.

File: src/components/application_manager/src/app_service_manager.cc (hunk 4 above)
Comment (ShobhitAd) [reply src/components/application_manager/src/app_service_manager.cc]: Resolved. Made changes to notification to include all published services, not just the updated one

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/on_system_capability_updated_notification.cc (hunk 1 above)
Comment (ShobhitAd) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/on_system_capability_updated_notification.cc]: Understood. I will remove the TODO comment

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/on_system_request_notification.cc [hunk 7]
```diff
@@ -106,17 +108,33 @@ void OnSystemRequestNotification::Run() {
     }
   }
 
-  if (mobile_apis::RequestType::PROPRIETARY == request_type) {
-    /* According to requirements:
-       "If the requestType = PROPRIETARY, add to mobile API fileType = JSON
-        If the requestType = HTTP, add to mobile API fileType = BINARY"
-       Also in Genivi SDL we don't save the PT to file - we put it directly in
-       binary_data */
+  LOG4CXX_DEBUG(logger_,
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/on_system_request_notification.cc]: The base branch was initially develop since it was before we had a base branch for the app services. Working on resolving the issue

File: src/components/application_manager/rpc_plugins/app_service_rpc_plugin/include/app_service_rpc_plugin/commands/hmi/on_as_system_capability_updated_notification.h [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+namespace app_service_rpc_plugin {
+namespace app_mngr = application_manager;
+
+namespace commands {
+
+/**
+ * @brief OnASSystemCapabilityUpdatedNotification command class
+ **/
+class OnASSystemCapabilityUpdatedNotification
```
Comment (JackLivio) [inline src/components/application_manager/rpc_plugins/app_service_rpc_plugin/include/app_service_rpc_plugin/commands/hmi/on_as_system_capability_updated_notification.h]: Should this class have "AS" in the name or be in the app service plugin? I see there was a SystemCapability interface created which is separate to the AppService interface. 

I would have imagined that the OnSystemCapabilityUpdated notification would have been included in the basic communication interface and this class was placed with the sdl_rpc plugin.

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/get_system_capability_request.cc [hunk 9]
```diff
@@ -128,10 +131,70 @@ void GetSystemCapabilityRequest::Run() {
...
+        mobile_apis::AppServiceType::eType service_type =
+            static_cast<mobile_apis::AppServiceType::eType>(
+                record[strings::service_manifest][strings::service_type]
+                    .asUInt());
+        service_types.insert(service_type);
+
+        // APP SERVICES
+        smart_objects::SmartObject app_services_capabilities(
+            smart_objects::SmartType_Map);
+        app_services_capabilities[strings::updated_app_service_record] = record;
```
Comment (JackLivio) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/get_system_capability_request.cc]: I think this is supposed to only include active service types. It looks like this could add multiple of the same service type.

Comment (JackLivio) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/get_system_capability_request.cc]: GetAllServiceRecords does not do a check for if the service is active.

Comment (ShobhitAd) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/get_system_capability_request.cc]: At the moment it adds all published services to the list of service records. Not sure if that is what we wanted for the OSCU notification and we can modify the behavior if required. 

Even though it will list multiple services of the same service type the service record contains a `serviceActive` param

File: src/components/application_manager/src/app_service_manager.cc [hunk 10]
```diff
@@ -106,4 +149,91 @@ std::vector<smart_objects::SmartObject> AppServiceManager::GetAllServices() {
...
+  smart_objects::SmartObject message(smart_objects::SmartType_Map);
+
+  message[strings::params][strings::message_type] = MessageType::kNotification;
+  message[strings::params][strings::function_id] =
+      mobile_apis::FunctionID::OnSystemCapabilityUpdatedID;
+
+  smart_objects::SmartObject system_capability =
+      smart_objects::SmartObject(smart_objects::SmartType_Map);
+  system_capability[strings::system_capability_type] =
+      mobile_apis::SystemCapabilityType::APP_SERVICES;
```
Comment (JackLivio) [inline src/components/application_manager/src/app_service_manager.cc]: Future Note: There might be other types of capabilities that can be updated other than app services. For example publishing a nav service could update the NAVIGATION system capability type in addition to the APP_SERVICE type. Lets confirm this in the evolution process during the NAVIGATION type review. Hold off on making the change for now.

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/extensions/system_capability_app_extension.cc [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+SystemCapabilityAppExtension::SystemCapabilityAppExtension(
+    sdl_rpc_plugin::SDLRPCPlugin& plugin, app_mngr_::Application& app)
+    : app_mngr_::AppExtension(
+          SystemCapabilityAppExtension::SystemCapabilityAppExtensionUID)
+    , plugin_(plugin)
+    , app_(app) {}
+
+SystemCapabilityAppExtension::~SystemCapabilityAppExtension() {}
+
+bool SystemCapabilityAppExtension::subscribeTo(
```
Comment (jacobkeeler) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/extensions/system_capability_app_extension.cc]: Class functions should be named with `UpperCamelCase`: https://google.github.io/styleguide/cppguide.html#Function_Names

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/get_system_capability_request.cc [hunk 12]
```diff
@@ -128,10 +131,70 @@ void GetSystemCapabilityRequest::Run() {
...
+    case mobile_apis::SystemCapabilityType::APP_SERVICES: {
+      smart_objects::SmartObject app_service_capabilities(
+          smart_objects::SmartType_Map);
+      smart_objects::SmartObject supported_types(
+          smart_objects::SmartType_Array);
+      smart_objects::SmartObject app_services(smart_objects::SmartType_Array);
+
+      std::vector<smart_objects::SmartObject> service_records =
+          application_manager_.GetAppServiceManager().GetAllServices();
+      std::set<mobile_apis::AppServiceType::eType> service_types;
```
Comment (ShobhitAd) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/get_system_capability_request.cc]: Change from `AppServiceType` to String

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/get_system_capability_request.cc [hunk 13]
```diff
@@ -128,10 +131,70 @@ void GetSystemCapabilityRequest::Run() {
...
+
+      for (auto& record : service_records) {
+        // SUPPORTED TYPES
+        mobile_apis::AppServiceType::eType service_type =
```
Comment (ShobhitAd) [inline src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/get_system_capability_request.cc]: Change from `AppServiceType` to String

File: src/components/application_manager/src/message_helper/message_helper.cc [hunk 14]
```diff
@@ -327,6 +327,73 @@ smart_objects::SmartObjectSPtr MessageHelper::CreateMessageForHMI(
...
+  system_capability[strings::system_capability_type] =
+      mobile_apis::SystemCapabilityType::APP_SERVICES;
+
+  smart_objects::SmartObject app_service_capabilities(
+      smart_objects::SmartType_Map);
+  smart_objects::SmartObject supported_types(smart_objects::SmartType_Array);
+  smart_objects::SmartObject app_services(smart_objects::SmartType_Array);
+
+  std::vector<smart_objects::SmartObject> service_records = all_services;
+  std::set<mobile_apis::AppServiceType::eType> service_types;
```
Comment (ShobhitAd) [inline src/components/application_manager/src/message_helper/message_helper.cc]: Change from `AppServiceType` to String

File: src/components/interfaces/HMI_API.xml [hunk 15]
```diff
@@ -3634,8 +3625,64 @@
...
+            <description> Only included in OnSystemCapabilityUpdated. Update reason for service record.</description>
+        </param>
+        <param name="updatedAppServiceRecord" type="AppServiceRecord" mandatory="true">
+            <description>Service record for a specific app service provider</description>
+        </param>
+    </struct>
+
+    <struct name="AppServicesCapabilities" since="5.1">
+        <description>Capabilities of app services including what service types are supported and the current state of services.</description>
+        <param name="servicesSupported" type="AppServiceType" array="true" mandatory="true">
```
Comment (ShobhitAd) [inline src/components/interfaces/HMI_API.xml]: Change from `AppServiceType` to String

File: src/components/interfaces/MOBILE_API.xml [hunk 16]
```diff
@@ -3973,8 +3957,68 @@
...
+            <description> Only included in OnSystemCapabilityUpdated. Update reason for service record.</description>
+        </param>
+        <param name="updatedAppServiceRecord" type="AppServiceRecord" mandatory="true">
+            <description>Service record for a specific app service provider</description>
+        </param>
+    </struct>
+
+    <struct name="AppServicesCapabilities" since="5.1">
+        <description>Capabilities of app services including what service types are supported and the current state of services.</description>
+        <param name="servicesSupported" type="AppServiceType" array="true" mandatory="true">
```
Comment (ShobhitAd) [inline src/components/interfaces/MOBILE_API.xml]: Change from `AppServiceType` to String

File: src/components/interfaces/HMI_API.xml [hunk 17]
```diff
@@ -3634,8 +3625,64 @@
...
+            <description>An array of supported app service types</description>
+        </param>
+        <param name="appServices" type="AppServiceCapability" array="true" mandatory="false">
+            <description>An array of currently available services. If this is an update to the capability the affected services will include an update reason in that item</description>
+        </param>
+    </struct>
+
     <!-- End App Services -->
 
+  <struct name="SystemCapabilities">
```
[the remark this answers is not in the document]
Comment (jacobkeeler) [reply src/components/interfaces/HMI_API.xml]: Fixed this myself in https://github.com/smartdevicelink/sdl_core/pull/2800/commits/cd8482bff603f0dbe841a4bd4b21fbf79ab4e880

File: src/components/interfaces/MOBILE_API.xml (hunk 16 above)
Comment (jacobkeeler) [reply src/components/interfaces/MOBILE_API.xml]: This is handled in the base branch, changes should be merged

File: src/components/interfaces/HMI_API.xml (hunk 15 above)
Comment (jacobkeeler) [reply src/components/interfaces/HMI_API.xml]: This is handled in the base branch, changes should be merged

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/extensions/system_capability_app_extension.h [hunk 18]
```diff
@@ -0,0 +1,83 @@
...
+
+  SystemCapabilitySubscriptions Subscriptions();
+
+  void SaveResumptionData(ns_smart_device_link::ns_smart_objects::SmartObject&
+                              resumption_data) OVERRIDE;
+
+  void ProcessResumption(
+      const smart_objects::SmartObject& resumption_data) OVERRIDE;
+
+  static unsigned SystemCapabilityAppExtensionUID;
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/extensions/system_capability_app_extension.h]: I was basing it off the UID in the vehicle_info_app_extension ``` /**
   * @brief VehicleInfoAppExtensionUID unique identifier of VehicleInfo
   * aplication extension
   */
  static unsigned VehicleInfoAppExtensionUID;```. I assign the value for the UID in the .cc file so setting it to const throws an error. I could try setting the value in the .h file itself I guess

File: src/components/interfaces/HMI_API.xml [hunk 19]
```diff
@@ -3634,8 +3625,64 @@
...
+        </param>
+        <param name="remoteControlCapability" type="RemoteControlCapabilities" mandatory="false">
+            <description>Describes extended capabilities of the module's phone feature</description>
+        </param>
+        <param name="appServicesCapabilities" type="AppServicesCapabilities" mandatory="false">
+            <description>An array of currently available services. If this is an update to the capability the affected services will include an update reason in that item</description>
+        </param>      
+  </struct>
+
+
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply src/components/interfaces/HMI_API.xml]: Resolved in https://github.com/smartdevicelink/sdl_core/pull/2800/commits/cd8482bff603f0dbe841a4bd4b21fbf79ab4e880#diff-f81183d34277b39fa042fd7e4eed256fR3693

File: src/components/application_manager/src/app_service_manager.cc [hunk 20]
```diff
@@ -81,19 +82,73 @@ smart_objects::SmartObject AppServiceManager::PublishAppService(
...
+              service_id,
+              mobile_apis::ServiceUpdateReason::REMOVED));
+  app_manager_.GetRPCService().ManageHMICommand(hmi_notification);
+
+  published_services_.erase(it);
+
+  return true;
+}
+
+void AppServiceManager::UnpublishAppServices(const uint32_t connection_key) {
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply src/components/application_manager/src/app_service_manager.cc]: I am uncertain as to whether this function would only be used by the `ApplicationManagerImpl::UnregisterApplication` function. At the moment we are only removing all the published services of an app when it unregisters but there might be other cases where we might have to do that. However to avoid confusion with `UnpublishAppService` I am changing it to `UnpublishServices`

File: src/components/application_manager/rpc_plugins/app_service_rpc_plugin/CMakeLists.txt [hunk 21]
```diff
@@ -28,7 +28,10 @@
 # ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 # POSSIBILITY OF SUCH DAMAGE.
 
-include_directories(include)
+include_directories(
+  include
+  ${COMPONENTS_DIR}/application_manager/rpc_plugins/sdl_rpc_plugin/include
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply src/components/application_manager/rpc_plugins/app_service_rpc_plugin/CMakeLists.txt]: Not sure if theres a better way to handle this but it is used to handle subscribing the published service to OSCU notifications. I was unable to put this in the ASM since extracting the app extension requires an app id, which is not passed in to the `PublishAppService` function(besides the HMI does not have one)

File: src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/extensions/system_capability_app_extension.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (jacobkeeler) [reply src/components/application_manager/rpc_plugins/sdl_rpc_plugin/include/sdl_rpc_plugin/extensions/system_capability_app_extension.h]: Hmmmm, adding `extern` might do it (like is done in `smart_object_keys.h`):
```
static extern const AppExtensionUID SystemCapabilityAppExtensionUID;
```
Regardless, the type should at least be `AppExtensionID` rather than just `unsigned`.
